docs/extended-tour.md

```diff
@@ -0,0 +1,159 @@
+# Extended Tour
+Welcome to the extended tour of node-cpp-skel. This documentation is especially handy if you're just starting out with building [Node Addon-ons](https://github.com/mapbox/cpp/blob/master/node-cpp.md) or writing C++ :tada:
+
+## Table of Contents:
+- [Walkthrough of example code](extended-tour.md#walkthrough-example-code)
+- [What does "build" mean?](extended-tour.md#what-does-build-mean)
+- [Software build components](extended-tour.md#software-build-components)
+- [Configuration files](extended-tour.md#configuration-files)
+- [Autogenerated files](extended-tour.md#autogenerated-files)
+- [Overall flow](extended-tour.md#overall-flow)
+
+# Walkthrough example code
+
+This skeleton includes a few examples of how you might design your application, including standalone functions and creating Objects/Classes. Both the synchronous and asynchronous versions of each are included to give an iterative example of what the progression from sync to async looks like. Let's run through reasons why you'd design your code in these ways:
+1. [Standalone function](API.md#hello-1)
+2. [Standalone asynchronous function](../API.md#hello_async)
+3. [Object/Class](../API.md#helloobject)
+4. [Asynchronous Object/Class](../API.md#helloobjectasync)
+
+
+### When would you use a standalone function?
+Standalone functions are simple and useful when your code is doing some kind of basic operation. They avoid the need for a class, since creating a class in this situation would add unnecessary complexity. Creating an asynchronous standalone function can be powerful if your code has a specific, yet expensive operation to do and would benefit from the use of multiple threads at one time to do that expensive work.
```

Comment:
> Relevant to mention anything about performance benefits by using standalone functions instead of object

I'm not sure, but great point that this is not totally clear. Can I leave it up to you to figure out how to integrate? I'll offer some more detailed thinking to further flesh out why we designed the way we did in the shaver..... the key to performance is efficiency (so reducing unneeded work - https://github.com/mapbox/cpp/blob/master/glossary.md#efficiency). Therefore, in vt-shaver we:

  - Created a class (`Filter` object) to hold parsed filters because the parsing step (converting a `string` into in-memory C++ objects) took time. We did not want to just pass a `string` object into the `shave` function, so we needed an custom object (a class) to hold the in-memory C++ objects representing the parsed filters in such a way that they could be accessed in a zero-copy way in the threadpool. This optimization worked on the assumption that `Filter` objects could be created once per style and passed per tile when shaving (to avoid needing to parse the filters string per tile). This is what happens in the `shave` function: we pass the `Filters` object and in the threadpool we access its memory to get the actually C++ filters.

We could have added a `shave` function on the `Filters` class. This would have had identical performance to the standalone function we went with. So the reason for the standalone function was, in this case, to separate concerns (https://en.wikipedia.org/wiki/Separation_of_concerns): Filters don't need to know about vector tiles so having a method involving vector tiles on the `Filter` object would have been messy as it combined concerns. They could be separated, so we separated them. So, the `shave` function was standalone because it could be. The key to performance was making the argument a custom class (`Filter` object) to hold pre-processed data needed for shaving.